Fails with assignment expression used in "any()" function

<!--
Thank you for taking the time to report an issue! We're glad to have you involved with Ruff.

If you're filing a bug report, please consider including the following information:

* A minimal code snippet that reproduces the bug.
* The command you invoked (e.g., `ruff /path/to/file.py --fix`), ideally including the `--isolated` flag.
* The current Ruff settings (any relevant sections from your `pyproject.toml`).
* The current Ruff version (`ruff --version`).
-->
Content of "test.py":
```python
if any((key := x) for x in ["ok"]):
    print(key)
```
Ruff command:
```
ruff --isolated test.py
```

Ruff output:
```
test.py:2:11: F821 Undefined name `key`
```

Ruff version: 0.0.261

--- a/crates/ruff/resources/test/fixtures/pyflakes/F821_0.py
+++ b/crates/ruff/resources/test/fixtures/pyflakes/F821_0.py
@@ -132,3 +132,8 @@ def in_ipython_notebook() -> bool:
     except NameError:
         return False  # not in notebook
     return True
+
+
+def named_expr():
+    if any((key := (value := x)) for x in ["ok"]):
+        print(key)

--- a/crates/ruff/resources/test/fixtures/pyflakes/F841_0.py
+++ b/crates/ruff/resources/test/fixtures/pyflakes/F841_0.py
@@ -121,3 +121,8 @@ def f(x: int):
             print("A")
         case y:
             pass
+
+
+def f():
+    if any((key := (value := x)) for x in ["ok"]):
+        print(key)

--- a/crates/ruff/src/checkers/ast/mod.rs
+++ b/crates/ruff/src/checkers/ast/mod.rs
@@ -4266,7 +4266,20 @@ impl<'a> Checker<'a> {
             }
         }
 
-        let scope = self.ctx.scope();
+        // Per [PEP 572](https://peps.python.org/pep-0572/#scope-of-the-target), named
+        // expressions in generators and comprehensions bind to the parent scope.
+        let scope_id = if binding.kind.is_named_expr_assignment() {
+            self.ctx
+                .scopes
+                .ancestor_scopes(self.ctx.scope_id)
+                .find(|scope| !scope.kind.is_generator())
+                .expect("Every scope must descend from the global scope.")
+                .id
+        } else {
+            self.ctx.scope_id
+        };
+
+        let scope = &self.ctx.scopes[scope_id];
         let binding = if let Some(index) = scope.get(name) {
             let existing = &self.ctx.bindings[*index];
             match &existing.kind {
@@ -4298,11 +4311,16 @@ impl<'a> Checker<'a> {
 
         // Don't treat annotations as assignments if there is an existing value
         // in scope.
-        let scope = self.ctx.scope_mut();
-        if !(binding.kind.is_annotation() && scope.defines(name)) {
-            scope.add(name, binding_id);
+        if binding.kind.is_annotation() && scope.defines(name) {
+            self.ctx.bindings.push(binding);
+            return;
         }
 
+        // Add the binding to the scope.
+        let scope = &mut self.ctx.scopes[scope_id];
+        scope.add(name, binding_id);
+
+        // Add the binding to the arena.
         self.ctx.bindings.push(binding);
     }
 
@@ -4579,9 +4597,10 @@ impl<'a> Checker<'a> {
             return;
         }
 
-        let current = self.ctx.scope();
+        let scope = self.ctx.scope();
+
         if id == "__all__"
-            && matches!(current.kind, ScopeKind::Module)
+            && scope.kind.is_module()
             && matches!(
                 parent.node,
                 StmtKind::Assign { .. } | StmtKind::AugAssign { .. } | StmtKind::AnnAssign { .. }
@@ -4619,7 +4638,7 @@ impl<'a> Checker<'a> {
 
                     // Grab the existing bound __all__ values.
                     if let StmtKind::AugAssign { .. } = &parent.node {
-                        if let Some(index) = current.get("__all__") {
+                        if let Some(index) = scope.get("__all__") {
                             if let BindingKind::Export(Export { names: existing }) =
                                 &self.ctx.bindings[*index].kind
                             {
@@ -4660,6 +4679,27 @@ impl<'a> Checker<'a> {
                 );
                 return;
             }
+        }
+
+        if self
+            .ctx
+            .expr_ancestors()
+            .any(|expr| matches!(expr.node, ExprKind::NamedExpr { .. }))
+        {
+            self.add_binding(
+                id,
+                Binding {
+                    kind: BindingKind::NamedExprAssignment,
+                    runtime_usage: None,
+                    synthetic_usage: None,
+                    typing_usage: None,
+                    range: expr.range(),
+                    source: Some(*self.ctx.current_stmt()),
+                    context: self.ctx.execution_context(),
+                    exceptions: self.ctx.exceptions(),
+                },
+            );
+            return;
         }
 
         self.add_binding(

--- a/crates/ruff/src/checkers/ast/mod.rs
+++ b/crates/ruff/src/checkers/ast/mod.rs
@@ -4267,19 +4267,19 @@ impl<'a> Checker<'a> {
         }
 
         // Per [PEP 572](https://peps.python.org/pep-0572/#scope-of-the-target), named
-        // expressions in generators and comprehensions bind to the parent scope.
+        // expressions in generators and comprehensions bind to the scope that contains the
+        // outermost comprehension.
         let scope_id = if binding.kind.is_named_expr_assignment() {
             self.ctx
                 .scopes
-                .ancestor_scopes(self.ctx.scope_id)
-                .find(|scope| !scope.kind.is_generator())
-                .expect("Every scope must descend from the global scope.")
-                .id
+                .ancestor_ids(self.ctx.scope_id)
+                .find_or_last(|scope_id| !self.ctx.scopes[*scope_id].kind.is_generator())
+                .unwrap_or(self.ctx.scope_id)
         } else {
             self.ctx.scope_id
         };
+        let scope = &mut self.ctx.scopes[scope_id];
 
-        let scope = &self.ctx.scopes[scope_id];
         let binding = if let Some(index) = scope.get(name) {
             let existing = &self.ctx.bindings[*index];
             match &existing.kind {
@@ -4317,7 +4317,6 @@ impl<'a> Checker<'a> {
         }
 
         // Add the binding to the scope.
-        let scope = &mut self.ctx.scopes[scope_id];
         scope.add(name, binding_id);
 
         // Add the binding to the arena.

--- a/crates/ruff/src/rules/pyflakes/rules/unused_variable.rs
+++ b/crates/ruff/src/rules/pyflakes/rules/unused_variable.rs
@@ -322,7 +322,7 @@ pub fn unused_variable(checker: &mut Checker, scope: ScopeId) {
         .map(|(name, index)| (name, &checker.ctx.bindings[*index]))
     {
         if !binding.used()
-            && binding.kind.is_assignment()
+            && (binding.kind.is_assignment() || binding.kind.is_named_expr_assignment())
             && !checker.settings.dummy_variable_rgx.is_match(name)
             && name != &"__tracebackhide__"
             && name != &"__traceback_info__"

--- a/crates/ruff_python_semantic/src/binding.rs
+++ b/crates/ruff_python_semantic/src/binding.rs
@@ -253,6 +253,7 @@ pub enum BindingKind<'a> {
     Annotation,
     Argument,
     Assignment,
+    NamedExprAssignment,
     Binding,
     LoopVar,
     Global,

